### PR TITLE
Javascript Modules and Bundling Guideline

### PR DESCRIPTION
**Pull request by alex40724**
This PR is a direct follup-up of #2477 and a general follow-up of #2445.

We discussed #2477 in a virtual meeting and a general wish was to find solutions for modularisation and bundling in one step. This new PR includes proposals for both issues.

The main ideas are

- Switch to ES6 modules: https://github.com/ILIAS-eLearning/ILIAS/pull/2746/files#diff-735e9dbac4f0c3ee7adb1352d14d6e6b
- Use [rollup.js](https://rollupjs.org/) to create bundles: https://github.com/ILIAS-eLearning/ILIAS/pull/2746/files#diff-c56fd1024190d495695c27a53546b62c

Why rollup.js and not webpack?

The main idea of this proposal is to move to a Javascript architecture based on ES6 modules on all levels. But Webpack does not allow to output ES6 modules, see [1], [2]. (a future version might do it). Using rollup.js on the other hand allows us to created bundles optionally on all levels and decide whether we a) resolve imports by creating bundles or b) let the browser resolve the imports.

There are several articles on the web that compare webpack and rollup, e.g. [3].

Please add your comments and thoughts in the PR. You are also invited to join the ILIAS Offline Group [4]. The group will discuss the PR in an upcoming virtual meeting.

[1] https://github.com/webpack/webpack/issues/2933
[2] https://github.com/webpack/webpack/issues/10958
[3] https://medium.com/webpack/webpack-and-rollup-the-same-but-different-a41ad427058c
[4] https://docu.ilias.de/goto_docu_grp_8190.html

### CODE REVIEW

File: docs/development/js/js-modules.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+- The sequence of script-tags is error-prone and relies on addJavascript calls in lots of different PHP components.
+- Ajax calls may result in missing dependencies to unavailable js files.
+- The approach does not allow to bundle and minify at least the core Javascript code.
+
+## Proposal
+
+### Use ES6 modules
+
+Modules SHOULD be implemented as ES6 modules. They should export their public API part using ES6 export statements.
```
Comment (Amstutz) [inline docs/development/js/js-modules.md:59]: Maybe use SHOULD here.

**Comment by Amstutz** [comment]
Hi @akill, thx a lot for working out this proposal. I really like the direction this is heading towards. IMO this is looking good so far. Our client side architecture really those need this level up in structure.

**Comment by alex40724** [comment]
I put this on the JF agenda just to make a short announcement.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 08 JUN 2020 : Alexander gave a short presentation about his guideline for javascript modules and bundling. He asks all developers for feedback and questions to finalise the PR. Questions can be asked in the SIG Offline: https://docu.ilias.de/goto_docu_grp_8190.html

**Comment by matthiaskunkel** [comment]
Jour Fixe, 22 JUN 2020 : We highly appreciate this suggestion and accept the PR for trunk. Please merge.

**Pull request merged**